File: Lexical.rb
# Rodrigo Fuentes Hernandez
# aka Blurred_9L
# Compiladores CUCEI 2013B
# Lexical.rb

require "./Token.rb"
require "./FiniteAtmt.rb"

class Lexical
    attr_reader :line, :curChar, :curLine
    
    private
        def findStart()
            keepGoing = true
            
            while keepGoing and @automata.endOfToken?( @line[@curChar] )
                @curChar += 1
                if curChar > @line.length()
                    if not nextLine()
                        keepGoing = false
                    end
                end
            end
            
            return keepGoing
        end
        
        def nextLine()
            success = false
            
            @curLine += 1
            while @curLine < @lines.length() and not success
                @line = @lines[@curLine]
                if not @line.empty?
                    success = true
                else
                    @curLine += 1
                end
            end
            @curChar = 0
            
            return success
        end
    
    public
        def initialize( lines, atmt )
            @lines = lines
            @line = ""
            @curChar = 0
            @curLine = 0
            @automata = atmt
        end
        
        def nextToken()
            moreTokens = true
            token = nil
            tokenTypes = Token.getTokenTypes()
            
            if @line.empty?
                if @lines.length() > 0
                    @line = @lines[@curLine]
                    @curChar = 0
                else
                    moreTokens = false
                end
            elsif curChar >= @line.length()
                ok = true
                while curChar >= line.length() and ok
                    ok = nextLine()
                end
                if not ok
                    moreTokens = false
                elsif not findStart()
                    moreTokens = false
                end
            elsif not findStart()
                moreTokens = false
            end
            
            @automata.curChar = @curChar
            @automata.line = @line
            if moreTokens and not @line.empty?
                symbol = @automata.getTokenString()
                @curChar = @automata.curChar
                token = Token.new( symbol, @automata.getType() )
                if tokenTypes.isKeyword?( symbol )
                    token.type = tokenTypes.types[symbol]
                end
            end
            
            return token
        end
end

File: Lexical.py
# encoding: utf-8
# Rodrigo Fuentes Hernandez
# aka Blurred_9L
# Compiladores CUCEI 2013B
# Lexical.py

import Token
import FiniteAtmt

class Lexical( object ):
    
    def __init__( self, lines, atmt ):
        self.lines = lines
        self.line = ""
        self.curChar = 0
        self.curLine = 0
        self.automata = atmt
        
    def findStart( self ):
        keepGoing = True
        
        while keepGoing and self.automata.isEndOfToken( self.line[self.curChar] ):
            self.curChar += 1
            if self.curChar > len( self.line ):
                if not self.nextLine():
                    keepGoing = False
        
        return keepGoing
        
    def nextLine( self ):
        success = False
        
        self.curLine += 1
        while self.curLine < len( self.lines ) and not success:
            self.line = self.lines[self.curLine]
            if not self.line == "":
                success = True
            else:
                self.curLine += 1
        
        self.curChar = 0
        
        return success
        
    def nextToken( self ):
        moreTokens = True
        token = None
        tokenTypes = Token.getTokenTypes()
        
        if self.line == "":
            if len( self.lines ) > 0:
                self.line = self.lines[self.curLine]
                self.curChar = 0
            else:
                moreTokens = False
        elif self.curChar >= len( self.line ):
            ok = True
            while self.curChar >= len( self.line ) and ok:
                ok = self.nextLine()
            if not ok:
                moreTokens = False
            elif not self.findStart():
                moreTokens = False
        elif not self.findStart():
            moreTokens = False
            
        self.automata.curChar = self.curChar
        self.automata.line = self.line
        if moreTokens and not self.line == "":
            symbol = self.automata.getTokenString()
            self.curChar = self.automata.curChar
            token = Token.Token( symbol, self.automata.getType() )
            if tokenTypes.isKeyword( symbol ):
                token.nodeType = tokenTypes.types[symbol]
                
        return token
